include/boost/asio/windows_sspi/stream.hpp

```diff
@@ -119,14 +162,67 @@ template <typename NextLayer, typename ConstBufferSequence> struct async_write_i
                                              net::buffer(m_message),
                                              net::transfer_exactly(m_message.size()),
                                              std::move(self));
-      self.complete(ec, net::buffer_size(m_buffer));
+      self.complete(ec, m_size_encrypted);
     }
   }
 
 private:
   NextLayer& m_next_layer;
   ConstBufferSequence m_buffer;
-  std::shared_ptr<sspi_impl> m_sspi;
+  std::shared_ptr<sspi_impl> m_sspi_impl;
+  std::vector<char> m_message;
+  size_t m_size_encrypted{0};
+};
+
+// TODO: Move away from this file
+template <typename NextLayer, typename MutableBufferSequence> struct async_read_impl : boost::asio::coroutine {
+  async_read_impl(NextLayer& next_layer, const MutableBufferSequence& buffer, std::shared_ptr<sspi_impl> sspi_impl)
+    : m_next_layer(next_layer)
+    , m_buffer(buffer)
+    , m_sspi_impl(std::move(sspi_impl)) {
+  }
+
+  template <typename Self> void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
+    BOOST_ASIO_CORO_REENTER(*this) {
+      while(m_sspi_impl->decrypted_data.empty()) {
+        // TODO: Find some way to make the sspi_impl, the decrypt
+        // function or something else be responsible for keeping track
+        // of state and buffer(s)
+        m_message.resize(0x10000 - m_sspi_impl->encrypted_data.size());
+        BOOST_ASIO_CORO_YIELD net::async_read(m_next_layer,
+                                              net::buffer(m_message),
+                                              std::move(self));
+        if (ec && length == 0 && m_sspi_impl->decrypted_data.empty()) {
+          self.complete(ec, 0);
+          return;
+        }
+        m_message.resize(length);
+        auto sc = m_sspi_impl->decrypt(m_message);
+        m_message.clear();
+        if (sc == SEC_E_INCOMPLETE_MESSAGE) {
+          continue;
+        }
+        if (sc == SEC_I_CONTEXT_EXPIRED) {
+          self.complete(net::error::eof, 0);
+          return;
+        }
+        if (sc != SEC_E_OK) {
+          self.complete(error::make_error_code(sc), 0);
+          return;
+        }
+      }
+
+      std::size_t to_return = std::min(net::buffer_size(m_buffer), m_sspi_impl->decrypted_data.size());
+      net::buffer_copy(m_buffer, net::buffer(m_sspi_impl->decrypted_data, to_return));
```

Comment:
`buffer_copy` returns the number of bytes actually copied. Consider asserting that this number is what you expected.

Reply:
Good point. Guess it should be an assertion as it would be a programming error if I somehow messed up the call to `std::min` above (or whatever this might look like in the future).